source/fab/metrics.py

```diff
@@ -0,0 +1,214 @@
+
+    # the pipe connections for the collated metrics, once they've stopped being sent
+    _collated_metrics_recv_conn, _collated_metrics_send_conn = Pipe(duplex=False)
+
+    # start the receiving process
+    _metric_recv_process = Process(
+        target=_read_metric,
+        daemon=True,  # todo: test this thoroughly, manually
+    )
+    _metric_recv_process.start()
+
+
+def _read_metric():
+    """
+    Intended to run as a child process, reading metrics created by other child processes.
+
+    Reads from the metric pipe until the pipe is closed,
+    at which point it pushes the collated metrics onto the "collated metrics" pipe and ends.
+
+    """
+    # An example metric is the time taken to preprocess a file; metrics['preprocess c']['my_file.c']
+    metrics: Dict[str, Dict[str, float]] = defaultdict(dict)
+
+    # todo: can we do this better?
+    # we run in a subprocess, so we get a copy of _metric_send_conn before it closes.
+    # when the calling process finally does close it, we'll still have an open copy of it,
+    # so the connection will still be considered open!
+    # therefore we close *OUR* copy of it now.
+    _metric_send_conn.close()
+
+    logger.debug('read_metric: waiting for metrics')
+    num_recorded = 0
+    while True:
+        try:
+            metric = _metric_recv_conn.recv()
+        except EOFError:
+            break
+
+        group, name, value = metric
+        metrics[group][name] = value
+        num_recorded += 1
+
+    logger.debug(f"read_metric: recorded {num_recorded} metrics")
+
+    # Send the collated metrics.
+    # Note: We tried a 2-way pipe so we'd only need one, but it didn't work
+    # so we moved on and created this second pipe for the total metrics.
+    # Might be a simplification in there.
+    _collated_metrics_send_conn.send(metrics)
+    _collated_metrics_send_conn.close()
+
+
+def send_metric(group: str, name: str, value):
+    """
+    Pass a metric to the reader process.
+
+    """
+    _metric_send_conn.send([group, name, value])  # type: ignore
+
+
+def stop_metrics():
+    """
+    Close the metrics pipe and reader process.
+
+    Return the final collection of metrics.
+
+    """
+    global _metric_recv_conn, _metric_send_conn
+    global _collated_metrics_recv_conn, _collated_metrics_send_conn
+    global _metric_recv_process
+
+    # Close the metrics recording pipe.
+    # The metrics recording process will notice and finish,
+    # and send the total metrics to the "collated metrics" pipe, which it then closes.
+    _metric_send_conn.close()  # type: ignore
+    _metric_recv_process.join(1)  # type: ignore
+
+    # pull out the total metrics
+    metrics = _collated_metrics_recv_conn.recv()  # type: ignore
+
+    # set these to none so metrics can be initialised again
+    _metric_recv_conn = _metric_send_conn = _metric_recv_process = _collated_metrics_recv_conn = None
+
+    return metrics
+
+
+def metrics_summary(metrics, output_folder: Path):
```

Comment:
For the future, consider refactoring this to be more modular making it easier to turn different plots (and data sets) on and off. Also to add new ones and remove old ones.